### DISCUSSION TITLE
Issues with "UART Operations" test

### DISCUSSION THREAD

**Discussion by prasmussen65** [DevSkillsHQ/discusssions#54] [Help] [answered]
It seems to me that there might be some inconsistencies between the description of the UART and the registers it uses, and how your test software works.
1. I can get 'passed' on the initialization:
     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/

Suite: Suite_1
  Test: test of BRR Registry ...passed
  Test: test of TER Registry ...passed
  Test: test of RER Registry ...passed
  Test: test of IER Registry ...passed
  Test: test of TDR Registry ...passed
  Test: test of RDR Registry ...passed
  Test: test execution init time ...passed

2. However, I can't get it to transmit characters.
I can see that in "test/uart_test.c" in "void test_Hello_message_transmission(void)" it tries this:
  uart_config cfg = {BR_38400, PARITY_NONE, 1, 0, 1, 1, 1};
  init_uart(&cfg);  // Initialises for UART baudrate 38400,8,None,1,None
  send_uart("Hello");
  CU_ASSERT(mem_reg.TDR == 'H');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == 'e');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == 'l');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == 'l');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == 'o');

So, that means it attempts to access the address 'mem_reg.TDR' to compare with the expected character, that my code should have written, ie. from within send_uart(), and not overwritten with a new value until allowed.

From the description of the TSR register, I am expecting a change of Bit0 from 1 to 0, and Bit1 from 0 to 1 by the hardware, when the transmission is complete. I then expected the function isr_tx_complete(void) to be called to indicate that the character has been sent.

However, I can't see the changes to the TSR register in isr_tx_complete() or right after having called "stub_memwrite (UART_TDR, text[0] )" to examine the TSR register.

The CU_ASSERT() lines above are not waiting when testing for a new character value, but expecting "isr_tx_complete()" to wait for the update to be correct.

Please let me know what I have misunderstood, or if there is/are any issue(s) in your code or description?

Best regards,
Peter

**Comment by prasmussen65** [comment]
I don't know if it matters, but I can add that I am running locally on a Linuxmint system, using terminals to edit and run the code.
I have not yet uploaded anything to Github, as it is not yet in a state and to be uploaded.
I will prefer to upload it when all tests are passed.

**Comment by andikovacs** [comment, marked as answer]
Hello @prasmussen65,

I am the author of this code challenge. 
The TSR register description is valid for the real hardware but as we run unit tests in a simulated environment there is no hardware to manipulate the bits in the TSR register. What we do in the unit test is to simulate that the transmission have completed by calling the ISR. In the simulated environment (i.e. the unit test) there is no real transmission happening but writing to the TDR register is considered to trigger a successful transmission which then will call the isr_tx_complete() routine.
If the UART driver implements the send_uart() and the isr_tx_complete() routines the transmission should be possible. 

Please let me know if you need further clarifications.

Best regards,
Andrei

**Reply by prasmussen65** [reply]
Hi Andrei,
I came a bit further, but I'm now at the end of the last test with multiple writes using ABC and 123.

To me it seems that the ASSERTs you have in:

void test_multiple_messages_are_correctly_handled(void) {
  uart_config cfg = {BR_38400, PARITY_NONE, 1, 0, 1, 1, 1};
  init_uart(&cfg);  // Initialises for UART baudrate 38400,8,None,1,None
  send_uart("ABC");
  CU_ASSERT(mem_reg.TDR == 'A');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == 'B');
  isr_tx_complete();
  send_uart("123");
  CU_ASSERT(mem_reg.TDR == 'C');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == '1');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == '2');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == '3');
}

Are not executing in the right time/order, because the last 3 tells me that they have C, 1 and 2, when I can see that they are 1, 2 and 3.

I stopped attempting to use TSR because it doesn't seem to follow your description, due to the simulation and not true hardware.

**Reply by andikovacs** [reply]
Hello Peter,

The intention of this test is to check if the function send_uart is re-entrant, i.e. if it can be called successfully while one transmission is in progress without messing up the current transmission but still preserving the possibility to send messages at any time.
More precisely send_uart("123") is called while the transmission of the message "ABC" is in progress. So if this test is passing it means that the send_uart function is not blocking the CPU and that other actions can be performed while handling the UART communication asynchronously.  Please let me know if you need further clarifications.

**Comment by prasmussen65** [comment]
Hi Andrei,
Thanks for the clarification!I am on my way to a meeting, but I will look into it and get back to you if I still have questions.
Best regards,Peter

    On Thursday, December 7, 2023 at 03:10:45 PM GMT+1, andikovacs ***@***.***> wrote:  
 
 


Hello @prasmussen65,

I am the author of this code challenge.
The TSR register description is valid for the real hardware but as we run unit tests in a simulated environment there is no hardware to manipulate the bits in the TSR register. What we do in the unit test is to simulate that the transmission have completed by calling the ISR. In the simulated environment (i.e. the unit test) there is no real transmission happening but writing to the TDR register is considered to trigger a successful transmission which then will call the isr_tx_complete() routine.
If the UART driver implements the send_uart() and the isr_tx_complete() routines the transmission should be possible.

Please let me know if you need further clarifications.

Best regards,
Andrei

**Comment by prasmussen65** [comment]
Hi Andrei,
I get it that it should be re-entrant, but it doesn't seem to me that you are doing the ASSERT tests in the right order in test_multiple_messages_are_correctly_handled(void), because I think I have sent the characters in the right order using stub_memwrite(), but I still get the following result:
Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests     11     11     10      1        0
             asserts     20     20     17      3      n/a

Best regards,Peter

    On Monday, December 11, 2023 at 03:21:12 PM GMT+1, andikovacs ***@***.***> wrote:  
 
 


Hello Peter,

The intention of this test is to check if the function send_uart is re-entrant, i.e. if it can be called successfully while one transmission is in progress without messing up the current transmission but still preserving the possibility to send messages at any time.
More precisely send_uart("123") is called while the transmission of the message "ABC" is in progress. So if this test is passing it means that the send_uart function is not blocking the CPU and that other actions can be performed while handling the UART communication asynchronously. Please let me know if you need further clarifications.

**Reply by andikovacs** [reply]
Maybe I misunderstood  the problem. Could you please share the output of the assertion which fails first? According to you what should be the correct order of the assertions?

**Comment by prasmussen65** [comment]
This is what I get:
[100%] Built target uart_test


     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/


Suite: Suite_1
  Test: test of BRR Registry ...passed
  Test: test of TER Registry ...passed
  Test: test of RER Registry ...passed
  Test: test of IER Registry ...passed
  Test: test of TDR Registry ...passed
  Test: test of RDR Registry ...passed
  Test: test execution init time ...passed
  Test: test of Hello message transmission ...passed
  Test: test of interrupt flag cleaning ...passed
  Test: test of multiple sends ...FAILED
    1. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:142  - mem_reg.TDR == 'C'
    2. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:144  - mem_reg.TDR == '1'
    3. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:146  - mem_reg.TDR == '2'
  Test: test execution tx isr time ...passed

Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests     11     11     10      1        0
             asserts     20     20     17      3      n/a

Elapsed time =    0.000 seconds
***@***.***:~/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main$
In test_multiple_messages_are_correctly_handled()  you have these lines:  send_uart("123");
  CU_ASSERT(mem_reg.TDR == 'C');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == '1');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == '2');
  isr_tx_complete();
  CU_ASSERT(mem_reg.TDR == '3');
}

I do a stub_memwrite (UART_TDR, output_character ) in isr_tx_complete() and I echo it to the display so I know it was written to TDR, but your ASSERT tells me it wasn't.
Please refresh my memory about how to merge my code, then you can get it all, including my (now commented out) debugging code.



    On Monday, December 11, 2023 at 03:46:41 PM GMT+1, andikovacs ***@***.***> wrote:  
 
 


Maybe I misunderstood the problem. Could you please share the output of the assertion which fails first? According to you what should be the correct order of the assertions?

**Reply by andikovacs** [reply]
From the output seems like the asserts for C, 1 and 2 are failing and this is most likely because something was overridden by the second call of send_uart. "My assert" is the standard CUnit assert used in the other tests as well so it should work as expected. Do you know if your stub prints the TDR value before the send_uart("123") function call or after?

**Comment by prasmussen65** [comment]
The 'C' character is sent to TDR before send_uart() receives '123' from the test code.
Uncommenting a few lines of debug lines provides this output:
Suite: Suite_1
  Test: test of BRR Registry ...passed
  Test: test of TER Registry ...passed
  Test: test of RER Registry ...passed
  Test: test of IER Registry ...passed
  Test: test of TDR Registry ...passed
  Test: test of RDR Registry ...passed
  Test: test execution init time ...passed
  Test: test of Hello message transmission ...
send_uart text = Hello 
isr_tx_complete: text_global_queue[queue_index=0][text_index[queue_index]=1] = 'e'
isr_tx_complete: text_global_queue[queue_index=0][text_index[queue_index]=2] = 'l'
isr_tx_complete: text_global_queue[queue_index=0][text_index[queue_index]=3] = 'l'
isr_tx_complete: text_global_queue[queue_index=0][text_index[queue_index]=4] = 'o'
passed
  Test: test of interrupt flag cleaning ...
send_uart text = X 
passed
  Test: test of multiple sends ...
send_uart text = ABC 
isr_tx_complete: text_global_queue[queue_index=2][text_index[queue_index]=1] = 'B'
isr_tx_complete: text_global_queue[queue_index=2][text_index[queue_index]=2] = 'C'

send_uart text = 123 
isr_tx_complete: text_global_queue[queue_index=3][text_index[queue_index]=1] = '2'
isr_tx_complete: text_global_queue[queue_index=3][text_index[queue_index]=2] = '3'
FAILED
    1. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:142  - mem_reg.TDR == 'C'
    2. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:144  - mem_reg.TDR == '1'
    3. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:146  - mem_reg.TDR == '2'
  Test: test execution tx isr time ...isr_tx_complete: text_global_queue[queue_index=5][text_index[queue_index]=5] = ''
FAILED
    1. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:158  - CU_ASSERT_TRUE(cpu_time_used < 0.000005)

Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests     11     11      9      2        0
             asserts     20     20     16      4      n/a

    On Monday, December 11, 2023 at 04:05:32 PM GMT+1, andikovacs ***@***.***> wrote:  
 
 


From the output seems like the asserts for C, 1 and 2 are failing and this is most likely because something was overridden by the second call of send_uart. "My assert" is the standard CUnit assert used in the other tests as well so it should work as expected. Do you know if your stub prints the TDR value before the send_uart("123") function call or after?

**Reply by andikovacs** [reply]
I reanalyzed the code and the test again and I still think that something is wrong with the software under test. I will try my best to explain with comments on the following test how things are supposed to happen:

void test_multiple_messages_are_correctly_handled(void) {
uart_config cfg = {BR_38400, PARITY_NONE, 1, 0, 1, 1, 1};
init_uart(&cfg); // Initialises for UART baudrate 38400,8,None,1,None
// At this point the UART is initialized and all registers are prepared for being able to send data.
send_uart("ABC");  // this function populates the TDR register with 'A'
CU_ASSERT(mem_reg.TDR == 'A');  // the test checks that the register was populated correctly
isr_tx_complete(); // the transmission of 'A' is successful and the ISR is called - this triggers also the transmission of 'B'
CU_ASSERT(mem_reg.TDR == 'B'); // the test checks that the register was populated correctly
isr_tx_complete(); // the transmission of 'B' is successful and the ISR is called - this triggers also the transmission of 'C' 
                             // i.e. populates the TDR register with 'C'
send_uart("123"); // send uart with a different message is called while  the transmission of 'C' is in progress. As no ISR was called the TDR 
                             // register shall still contain 'C'
CU_ASSERT(mem_reg.TDR == 'C'); // The test checks that the register still has 'C' i.e. the call of send_uart do not mess up the 
                                                            // transmission in progress
isr_tx_complete(); // The transmission of 'C' successfully completes and the uart driver starts the transmission of the next character 
                             // which is '1'
CU_ASSERT(mem_reg.TDR == '1'); // the test checks that the register was populated correctly
// .... and so on, the same logic applies.
isr_tx_complete();
CU_ASSERT(mem_reg.TDR == '2');
isr_tx_complete();
CU_ASSERT(mem_reg.TDR == '3');
}

I cannot understand how in your example the TDR register can be populated with '1' before the transmission of 'C' is successful i.e. the ISR is executed by someone else then the unit test or the TDR register is overridden by the send_uart function.
On the real hardware the ISR is only called by the hardware and in the test environment the hardware is simulated by the unit test. If by mistake the ISR is called by the driver that is a mistake and the UART driver is not supposed to work like that.

**Comment by prasmussen65** [comment]
It is not obvious for me how to submit my file changes for this test, when reading explanations on Github.However, I have only changed the uart.c file and attached it here.
If you execute it with #define DEBUG active, then you will see that I am sending the characters in the right order, and the last test should pass.
[100%] Built target uart_test


     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/


Suite: Suite_1
  Test: test of BRR Registry ...passed
  Test: test of TER Registry ...passed
  Test: test of RER Registry ...passed
  Test: test of IER Registry ...passed
  Test: test of TDR Registry ...passed
  Test: test of RDR Registry ...passed
  Test: test execution init time ...passed
  Test: test of Hello message transmission ...
send_uart(), line 57 First character written (H) (full text = 'Hello') 
isr_tx_complete(), line70, Next character written = 'e'
isr_tx_complete(), line70, Next character written = 'l'
isr_tx_complete(), line70, Next character written = 'l'
isr_tx_complete(), line70, Next character written = 'o'
passed
  Test: test of interrupt flag cleaning ...
send_uart(), line 57 First character written (X) (full text = 'X') 
passed
  Test: test of multiple sends ...
send_uart(), line 57 First character written (A) (full text = 'ABC') 
isr_tx_complete(), line70, Next character written = 'B'
isr_tx_complete(), line70, Next character written = 'C'

send_uart(), line 57 First character written (1) (full text = '123') 
isr_tx_complete(), line70, Next character written = '2'
isr_tx_complete(), line70, Next character written = '3'
FAILED
    1. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:142  - mem_reg.TDR == 'C'
    2. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:144  - mem_reg.TDR == '1'
    3. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:146  - mem_reg.TDR == '2'
  Test: test execution tx isr time ...isr_tx_complete(), line70, Next character written = ''
FAILED
    1. /home/peter/Additude/UART_Level2_c/uart_level2_c_8e5ad13-e7d5eg-main/test/uart_test.c:158  - CU_ASSERT_TRUE(cpu_time_used < 0.000005)

Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests     11     11      9      2        0
             asserts     20     20     16      4      n/a

Elapsed time =    0.000 seconds


Everything is only written in order to make this test suite pass, not a general UART driver.
Best regards,Peter

    On Monday, December 11, 2023 at 08:44:54 PM GMT+1, andikovacs ***@***.***> wrote:  
 
 


I reanalyzed the code and the test again and I still think that something is wrong with the software under test. I will try my best to explain with comments on the following test how things are supposed to happen:

void test_multiple_messages_are_correctly_handled(void) {
uart_config cfg = {BR_38400, PARITY_NONE, 1, 0, 1, 1, 1};
init_uart(&cfg); // Initialises for UART baudrate 38400,8,None,1,None
// At this point the UART is initialized and all registers are prepared for being able to send data.
send_uart("ABC"); // this function populates the TDR register with 'A'
CU_ASSERT(mem_reg.TDR == 'A'); // the test checks that the register was populated correctly
isr_tx_complete(); // the transmission of 'A' is successful and the ISR is called - this triggers also the transmission of 'B'
CU_ASSERT(mem_reg.TDR == 'B'); // the test checks that the register was populated correctly
isr_tx_complete(); // the transmission of 'B' is successful and the ISR is called - this triggers also the transmission of 'C'
// i.e. populates the TDR register with 'C'
send_uart("123"); // send uart with a different message is called while the transmission of 'C' is in progress. As no ISR was called the TDR
// register shall still contain 'C'
CU_ASSERT(mem_reg.TDR == 'C'); // The test checks that the register still has 'C' i.e. the call of send_uart do not mess up the
// transmission in progress
isr_tx_complete(); // The transmission of 'C' successfully completes and the uart driver starts the transmission of the next character
// which is '1'
CU_ASSERT(mem_reg.TDR == '1'); // the test checks that the register was populated correctly
// .... and so on, the same logic applies.
isr_tx_complete();
CU_ASSERT(mem_reg.TDR == '2');
isr_tx_complete();
CU_ASSERT(mem_reg.TDR == '3');
}

I cannot understand how in your example the TDR register can be populated with '1' before the transmission of 'C' is successful i.e. the ISR is executed by someone else then the unit test or the TDR register is overridden by the send_uart function.
On the real hardware the ISR is only called by the hardware and in the test environment the hardware is simulated by the unit test. If by mistake the ISR is called by the driver that is a mistake and the UART driver is not supposed to work like that.

**Reply by andikovacs** [reply]
The problem is between the following lines:

isr_tx_complete(), line70, Next character written = 'C'

send_uart(), line 57 First character written (1) (full text = '123')

'1' should not be written until the transmission of 'C' is not complete. And transmission of 'C' is not complete prior to the ISR being executed. 

I cannot see the attached file. Pushing to a branch should work as usual in git but maybe @fenske can support.
Anyways we are not debugging the code now but discussing a principle, which is that 'C' should be in the TDR register until the ISR is executed and that is not happening. The send_uart function overwrites the register value too early with '1' and that causes the test to fail. The logs you are showing prove that 'C' was written to the TDR register but that is not sufficient. As said earlier 'C' should be in the TDR register until the HW (or the test) flags that the transmission was successfully completed, i.e. the ISR is executed.
